### PR TITLE
Refresh templates on dashboard load

### PR DESCRIPTION
**Pull request by 121watts**
- [x] CHANGELOG.md updated with a link to the PR (not the Issue)
  - [x] Rebased/mergable
  - [x] Tests pass
  - [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

Connect #1551 

### The problem
If the values of a template variable are updated those changes were never updated.

### The Solution
Update tempVars on dashboard mount.

### Notes
This made me sad but @cryptoquick has written PHP and he's actually totally fine with the way this looks

### CODE REVIEW

File: ui/src/dashboards/components/TemplateControlBar.js [hunk 1]
```diff
@@ -74,4 +112,4 @@ TemplateControlBar.propTypes = {
   isOpen: bool,
 }
 
-export default TemplateControlBar
+export default ControlBarWrapper
```
Comment (jademcgough) [inline ui/src/dashboards/components/TemplateControlBar.js]: This export is not being used anywhere, and in fact `TemplateControlBar` is still being imported in `Dashboard.js`. Did something get accidentally left out of this PR?

Comment (jademcgough) [inline ui/src/dashboards/components/TemplateControlBar.js]: Oh, nvm - I didn't realize that the default export could be given an arbitrary name.

File: ui/src/dashboards/components/TemplateControlBar.js [hunk 2]
```diff
@@ -57,6 +57,44 @@ const TemplateControlBar = ({
...
+ControlBarWrapper.propTypes = {
+  isOpen: bool,
+  source: shape().isRequired,
+  dashboard: shape().isRequired,
+  onSelectTemplate: func.isRequired,
+  updateTempVarValues: func.isRequired,
+  onOpenTemplateManager: func.isRequired,
+}
+
 TemplateControlBar.propTypes = {
```
Comment (jademcgough) [inline ui/src/dashboards/components/TemplateControlBar.js]: I'd prefer to move the `TemplateControlBar` propTypes up to line 60, underneath its component definition (and immediately after the `PropTypes` destructuring.

**Comment by cryptoquick** [comment]
The code in this PR is not yet ready. Upon further testing, there's a race condition where updated data for when the component mounts often loads earlier than when the dashboard data even loads.

Will have to research a remedy.

**Comment by 121watts** [comment]
@hunter repro steps?

**Comment by cryptoquick** [comment]
@121watts and I worked on a solution to this, but I think he had a different branch checked out or something, and I don't know how to use his setup very well, so we might have some git madness on this branch or sth

File: ui/src/dashboards/components/TemplateControlBar.js (hunk 2 above)
Comment (121watts) [inline ui/src/dashboards/components/TemplateControlBar.js]: @jademcgough we moved the async action from `TemplateControlBar` which means we no longer need the wrapper component :)

**Comment by jaredscheib** [comment]
What were the repro steps on this, so I can test it?

**Comment by jaredscheib** [comment]
The race condition was when `getDashboards()` resolved after `updateTempVars()`, which should now be impossible with this change.

File: ui/src/dashboards/actions/index.js [hunk 3]
```diff
@@ -123,12 +128,26 @@ export const templateVariableSelected = (dashboardID, templateID, values) => ({
...
+  },
+})
+
 // Async Action Creators
 
 export const getDashboardsAsync = () => async dispatch => {
   try {
     const {data: {dashboards}} = await getDashboardsAJAX()
     dispatch(loadDashboards(dashboards))
+    return dashboards
```
Comment (jaredscheib) [inline ui/src/dashboards/actions/index.js]: funky to see a dispatch and then a return in a thunk, but works fine and the alternative (purely redux-state-change-driven) is not necessarily better, and definitely more work.

File: ui/src/dashboards/containers/DashboardPage.js [hunk 4]
```diff
@@ -57,13 +58,23 @@ class DashboardPage extends Component {
...
+      source,
     } = this.props
 
-    getDashboardsAsync(dashboardID)
+    const dashboards = await getDashboardsAsync(dashboardID)
+    const dashboard = dashboards.find(d => d.id === +dashboardID)
+
+    // Refresh and persists template variable values on dashboard mount
+    await updateTempVarValues(source, dashboard)
+    await putDashboardByID(dashboardID)
```
Comment (jaredscheib) [inline ui/src/dashboards/containers/DashboardPage.js]: could you add a comment that explains this pattern?

File: ui/src/dashboards/actions/index.js [hunk 5]
```diff
@@ -145,6 +164,17 @@ export const putDashboard = dashboard => async dispatch => {
   }
 }
 
+export const putDashboardByID = dashboardID => async (dispatch, getState) => {
+  try {
+    const {dashboardUI: {dashboards}} = getState()
```
Comment (jaredscheib) [inline ui/src/dashboards/actions/index.js]: could you add a comment about why you're using `getState`, having to do with needing to not use a stale dashboard after `updateTempVarValues()`?

File: ui/src/dashboards/containers/DashboardPage.js [hunk 6]
```diff
@@ -57,13 +58,23 @@ class DashboardPage extends Component {
...
+      dashboardActions: {
+        getDashboardsAsync,
+        updateTempVarValues,
+        putDashboardByID,
+      },
+      source,
     } = this.props
 
-    getDashboardsAsync(dashboardID)
+    const dashboards = await getDashboardsAsync(dashboardID)
```
Comment (jaredscheib) [inline ui/src/dashboards/containers/DashboardPage.js]: also, while you're at it, this is old code, but `getDashboardsAsync` doesn't accept any arguments, so passing in `dashboardID` here is misleading

Comment (121watts) [inline ui/src/dashboards/containers/DashboardPage.js]: for sure!

**Pull request merged**